### PR TITLE
8344882: (bf) Temporary direct buffers should not count against the upper limit on direct buffer memory

### PR DESCRIPTION
**Pull request by bplb**
Make the memory used by internal temporary direct buffers not count towards the upper limit on direct buffer memory.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Change requires CSR request [JDK-8344997](https://bugs.openjdk.org/browse/JDK-8344997) to be approved
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8344882](https://bugs.openjdk.org/browse/JDK-8344882): (bf) Temporary direct buffers should not count against the upper limit on direct buffer memory (**Enhancement** - P4)
 * [JDK-8344997](https://bugs.openjdk.org/browse/JDK-8344997): (bf) Temporary direct buffers should not count against the upper limit on direct buffer memory (**CSR**)

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22339/head:pull/22339` \
`$ git checkout pull/22339`

Update a local copy of the PR: \
`$ git checkout pull/22339` \
`$ git pull https://git.openjdk.org/jdk.git pull/22339/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22339`

View PR using the GUI difftool: \
`$ git pr show -t 22339`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22339.diff">https://git.openjdk.org/jdk/pull/22339.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22339#issuecomment-2495109186)
</details>

### CODE REVIEW

**Comment by bplb** [comment]
This change should also fix [JDK-8340728](https://bugs.openjdk.org/browse/JDK-8340728) and [JDK-8342849](https://bugs.openjdk.org/browse/JDK-8342849) as a side effect.

File: src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template [hunk 1]
```diff
@@ -72,14 +74,18 @@ class Direct$Type$Buffer$RW$$BO$
     // Cached unaligned-access capability
     static final boolean UNALIGNED = Bits.unaligned();
 
-    private record Deallocator(long address, long size, int capacity) implements Runnable {
+    private @Stable boolean temporary; // defaults to false
```
Comment (ExE-Boss) [inline src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template:77]: Any reason this is `@Stable` instead of `final`?

Comment (bplb) [reply src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template:77]: I think it avoids setting it to `false` when it is not `true`.

**Comment by AlanBateman** [comment]
While you are there, can you also look into not creating a Cleaner for the temporary direct buffers? They are explicitly deallocated when they can't be returned to the TL cache or when the thread terminates (there is a hook in thread exit for this).

Also, on the surface it might look we could use a thread confined Arena and allocate MemorySegments and get a BB view but these temporary buffers are actually carrier local when there are virtual threads mounted.

**Comment by minborg** [comment]
Could we add some tests also?

**Comment by bplb** [comment]
> Could we add some tests also?

Do you have any suggestions? That the buffers do not affect the memory limit is implicitly tested by the tests associated with the issues linked in Jira.

**Comment by bplb** [comment]
> While you are there, can you also look into not creating a Cleaner for the temporary direct buffers? They are explicitly deallocated when they can't be returned to the TL cache or when the thread terminates (there is a hook in thread exit for this).

At present the explicit deallocation uses the cleaner:
```
    private static void free(ByteBuffer buf) {
        ((DirectBuffer)buf).cleaner().clean();
    } 
```

**Comment by AlanBateman** [comment]
> At present the explicit deallocation uses the cleaner:

Right but if we change this then you should find the "temporary" field goes away as it will be different deallocator.

File: src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template [hunk 2]
```diff
@@ -72,6 +74,8 @@ class Direct$Type$Buffer$RW$$BO$
     // Cached unaligned-access capability
     static final boolean UNALIGNED = Bits.unaligned();
 
+    private @Stable boolean temporary; // defaults to false
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template:77]: Is this used now?

File: src/java.base/share/classes/jdk/internal/access/JavaNioAccess.java [hunk 3]
```diff
@@ -67,6 +67,11 @@ public interface JavaNioAccess {
      */
     ByteBuffer newHeapByteBuffer(byte[] hb, int offset, int capacity, MemorySegment segment);
 
+    /**
+     * Used by {@code sun.nio.ch.Util}.
+     */
+    ByteBuffer allocateDirectTemporary(int cap);
```
Comment (AlanBateman) [inline src/java.base/share/classes/jdk/internal/access/JavaNioAccess.java:73]: In Util, the phrase used is "temporary buffer and the methods are named like getTemporaryDirectBuffer. The SharedSecrets transposes the words, don't know if that was intended or not.

File: src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template (hunk 2 above)
Comment (bplb) [reply src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template:77]: No. I removed it in a version I did not check in but forgot here.

**Comment by AlanBateman** [comment]
One thing we'll need to think about is observability. There may be an argument to track the usage as a buffer pool "temporary-direct" to add to the 3 pools that are currently exposed by BufferPoolMXBean.

File: src/java.base/share/classes/jdk/internal/access/JavaNioAccess.java (hunk 3 above)
Comment (bplb) [reply src/java.base/share/classes/jdk/internal/access/JavaNioAccess.java:73]: It was intended but I think they should match. I'll change the on in JavaNioAccess.

File: src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template (hunk 2 above)
Comment (bplb) [reply src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template]: Changed in [9d84aa9](https://github.com/openjdk/jdk/pull/22339/commits/9d84aa9cfd718265c72e26eff9434f8817d25d80).

File: src/java.base/share/classes/jdk/internal/access/JavaNioAccess.java (hunk 3 above)
Comment (bplb) [reply src/java.base/share/classes/jdk/internal/access/JavaNioAccess.java]: Changed in https://github.com/openjdk/jdk/pull/22339/commits/9d84aa9cfd718265c72e26eff9434f8817d25d80.

File: src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template [hunk 4]
```diff
@@ -39,7 +39,9 @@ import jdk.internal.misc.ScopedMemoryAccess.ScopedAccessError;
 import jdk.internal.misc.VM;
 import jdk.internal.ref.Cleaner;
 import sun.nio.ch.DirectBuffer;
-
+#if[byte]
+import jdk.internal.vm.annotation.Stable;
+#end[byte]
```
[the remark this answers is not in the document]
Comment (bplb) [reply src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template]: Removed in 4e4cdbb.

**Comment by AlanBateman** [comment]
I think you can simplify this more by allocating with Unsafe and use the DirectByteBuffer(int, long) constructor. That would avoid needing the DBB code to know anything about temporary buffers and also means the allocate/free with using Unsafe in one place rather than assume that DirectByteBuffer uses Unsafe. The SharedSecrets would change to newDirectBuffer.

**Comment by bplb** [comment]
> the DirectByteBuffer(int, long) constructor

I think you intend the constructor which is now
```
    private Direct$Type$Buffer(long addr, long cap) {
```
at line 172 in the latest commit.

**Comment by bplb** [comment]
> /label remove security

Thanks. I did not know why that was added.

**Comment by bplb** [comment]
> I think you can simplify this more by allocating with Unsafe and use the DirectByteBuffer(int, long) constructor.

Changed as suggested in 1d7588a.

**Review by AlanBateman — APPROVED** [review]
I think you've got this to a good place.

**Review by liach — COMMENTED** [review]
The latest revision does look much cleaner than previous iterations.

**Pull request closed without merging**